### ISSUE TITLE
Windows Support

### ISSUE THREAD

**Issue by StoneyJackson** [ourPLCC/plcc#80]
@jashelio @fosler 

Without a reasonably easy way manually and automatically test Windows batch files, we will need to strongly consider dropping support for Windows batch files.

I've been looking for ways to test Windows batch files in a GitHub workflow/action, but haven't found anything substantial. I really need to be able to fire up a virtual environment that is able to run batch as if it were in a Windows environment so that I can manually and automatically test the batch files. Without this we're going to have a hard time keeping the Windows side of things up to date and working.

Alternatively, we might want to direct Windows users toward Linux VMs and/or emulators. I know Windows now ships with WSL which allows one to install a any number of flavors of linux. Docker also provides a VM. So the Docker approach fits here too.

Thoughts? Suggestions?

**Comment by jashelio** [comment]
I looked around a bit myself and did not come up with any easy answer. As a longer term solution, does PowerShell come with all recent versions of Windows, and if so, would there be an easier way to run script files written for it?

Or is making students learn PowerShell too much of a burden? Considering my students’ lack of skill in bash, I don’t think it could be worse. But then again I have never used PowerShell.

Jim

**Comment by StoneyJackson** [comment]
@jashelio Good call on PowerShell. At first glance PowerShell looks promising. Here are some resources that I plan to look at when I circle back around.

* https://github.com/PowerShell/PowerShell
* https://devblogs.microsoft.com/scripting/use-docker-to-automate-testing-of-powershell-core-scripts/

**Comment by StoneyJackson** [comment]
So before we adopt another whole technology, I wanted to know how hard it is to get a Bash environment on Windows? Here's what I did.

1. Grabbed my wife's 64-bit Acer running Windows 11.
2. Ran Terminal as administrator
3. Typed `wsl --install` and waited until it was done (< 5m).
4. Rebooted and waited until it finished installing Ubuntu (< 5m).
5. In Terminal, selected the dropdown next to the tabs and selected Ubuntu, at which point I was in a bash shell.

This was a whole lot harder 20 years ago. If it's that easy now, I'm not sure we need to adopt PowerShell or continue to support Batch files.

I'm going to continue to install PLCC into that environment. I'll report back when I'm done.

**Comment by StoneyJackson** [comment]
1. Ran our new installer for bash
2. Update ~/.bashrc as described in the output of the installer.
3. Ran the tests given in the output of the installer and found Python 3.10 is already installed, but needed JDK... and it gave me the command to do it...
4. Ran `sudo apt update` and then `sudo apt install openjdk-17-jdk-headless`
5. Started a new shell.

PLCC is ready to go.

I'm not saying this is trivial. But I'm not convinced that PowerShell would make this easier.

Now what about VS Code integration... Be right back...

**Comment by StoneyJackson** [comment]
1. Installed VS Code into Windows.
2. Installed WSL extension into VS Code.
3. Opened a Terminal, opened an Ubuntu tab
4. Ran `code .` to start VS Code in the current directory, it installed some stuff, and launched VS Code.
5. Started a terminal inside VS Code, it's Bash running inside Ubuntu, and PLCC is working.

The only thing to address is getting files into and out of Ubuntu (filesystem stuff)... quick google search... Inside Ubuntu, run `explorer.exe .` and it opens a file browser opened to the current directory. Now drag stuff in and out as you like.

So I think I've convinced myself that we can reasonably consolidate to Bash. Although I'm open to arguments to the contrary.

Thoughts?
